File: src/RpgmDecrypt.Core/Crypto.fs
namespace RpgmDecrypt.Core

open System

/// Crypto primitives for the RPG Maker XOR scheme.
///
/// The MV and MZ engines share a single symmetric operation:
///
///     cipher_byte[i] = plain_byte[i] XOR key[i MOD keyLength]
///
/// The same byte stream XOR-applied against the same key produces the
/// plaintext; applying the XOR twice on a ciphertext gives back the
/// ciphertext (XOR is involutory). So we have one function, no separate
/// "decrypt key" — and the same call we test below proves it.
///
/// Specs / references (clean room):
///   * rpgmakerweb.com public docs: <https://rpgmakerweb.com>
///   * reverse-engineered algorithm publicly documented in many community
///     wikis; cross-checked against uuksu/RPGMakerDecrypter (MIT, public
///     source — used as a reference for byte ordering, never copy-pasted)
///     and Petschko/RPG-Maker-MV-Decrypter (MIT, archive-only).
module Crypto =

    let private ascii (s: string) : byte[] =
        System.Text.Encoding.ASCII.GetBytes s

    // ---- Magic-byte constants ---------------------------------------------
    let magicMvHeader     : byte[] = ascii "RPGMV"
    let magicMzHeader     : byte[] = ascii "RPGMZ"
    let magicRgssadPrefix : byte[] = ascii "RGSSAD\0" // 7 bytes
    // PNG header carries 0x89, outside ASCII's 7-bit range; we hard-code
    // the eight bytes so the result is byte-exact (Encoding.ASCII would
    // silently map 0x89 -> '?'). RFC 2083 PNG signature.
    let magicPng  : byte[] =
        [| 0x89uy; 0x50uy; 0x4Euy; 0x47uy; 0x0Duy; 0x0Auy; 0x1Auy; 0x0Auy |]
    let magicOgg  : byte[] = ascii "OggS"
    let magicM4a  : byte[] = ascii "ftyp"
    let magicRiff : byte[] = ascii "RIFF"
    let magicWebp : byte[] = ascii "WEBP"
    let magicJpg  : byte[] = [| 0xFFuy; 0xD8uy; 0xFFuy |]

    /// Proper equality helper for fixed-size substrings inside a larger array.
    let subArrayEq (offset: int) (length: int) (expected: byte[]) (arr: byte[]) : bool =
        if arr.Length < offset + length then false
        elif expected.Length < length then false
        else
            let mutable ok = true
            for i in 0 .. length - 1 do
                if arr.[offset + i] <> expected.[i] then ok <- false
            ok

    /// Does `arr` start with the bytes of `prefix`?
    let startsWith (prefix: byte[]) (arr: byte[]) : bool =
        if arr.Length < prefix.Length then false
        else subArrayEq 0 prefix.Length prefix arr

    /// Was the ciphertext already plaintext? True if `head` starts with a
    /// known unencrypted magic. Caller passes the first 16 bytes.
    let looksLikePlaintext (head: byte[]) : bool =
        if head.Length = 0 then false
        else
            startsWith magicPng head
            || startsWith magicOgg head
            || startsWith magicJpg head
            || (head.Length >= 12 && startsWith magicRiff head && subArrayEq 8 4 magicWebp head)
            || (head.Length >= 8 && subArrayEq 4 4 magicM4a head)

    /// Single hex character -> 0..15. Raises on bad char.
    let private hexNibble (c: char) : int =
        match c with
        | '0' -> 0  | '1' -> 1  | '2' -> 2  | '3' -> 3
        | '4' -> 4  | '5' -> 5  | '6' -> 6  | '7' -> 7
        | '8' -> 8  | '9' -> 9
        | 'a' -> 10 | 'b' -> 11 | 'c' -> 12 | 'd' -> 13
        | 'e' -> 14 | 'f' -> 15
        | _   -> invalidArg "hex" (sprintf "non-hex char '%c'" c)

    /// Decode a 32-char hex string into a 16-byte array. The MV/MZ key
    /// is always presented as this exact shape in System.json.
    let decodeHexKey (hex: string) : byte[] =
        let cleaned = hex.Trim().ToLowerInvariant()
        if cleaned.Length <> 32 then
            invalidArg "hex" (sprintf "encryption key must be 32 hex chars, got %d" cleaned.Length)
        let bytes = Array.zeroCreate<byte> 16
        for i in 0 .. 15 do
            let hi = hexNibble cleaned.[2 * i]
            let lo = hexNibble cleaned.[2 * i + 1]
            bytes.[i] <- byte ((hi <<< 4) ||| lo)
        bytes

    /// XOR `buf` against `key` cyclically. Symmetric — encrypt == decrypt.
    let xorTransform (key: byte[]) (buf: byte[]) : byte[] =
        if key.Length = 0 then
            invalidArg "key" "XOR key must be non-empty"
        let n = buf.Length
        let out = Array.zeroCreate<byte> n
        let klen = key.Length
        for i in 0 .. n - 1 do
            out.[i] <- buf.[i] ^^^ key.[i % klen]
        out

    let isMvMagicHeader (head: byte[]) : bool =
        head.Length >= magicMvHeader.Length && startsWith magicMvHeader head

    let isMzMagicHeader (head: byte[]) : bool =
        head.Length >= magicMzHeader.Length && startsWith magicMzHeader head

    let isRgssadMagic (head: byte[]) : bool =
        head.Length >= magicRgssadPrefix.Length && startsWith magicRgssadPrefix head

    let isZipMagic (head: byte[]) : bool =
        head.Length >= 4
        && head.[0] = byte 'P'
        && head.[1] = byte 'K'
        && head.[2] = 0x03uy
        && head.[3] = 0x04uy

    let zeroFill (buf: byte[]) : unit =
        if not (isNull buf) then
            for i in 0 .. buf.Length - 1 do buf.[i] <- 0uy

File: ocaml/lib/crypto.ml
(* Source — XOR scheme + magic-byte helpers + hex decode.
   F# byte[] -> OCaml bytes throughout. Clean-room reimplementation;
   same references as the F# version (rpgmakerweb.com docs, community wikis). *)

(* ---- Magic-byte constants (as bytes) ---------------------------------- *)
let magic_mv_header = Bytes.of_string "RPGMV"
let magic_mz_header = Bytes.of_string "RPGMZ"

(* 7 bytes "RGSSAD\0" — explicit so no escape-handling surprise. *)
let magic_rgssad_prefix =
  Bytes.of_string "\x52\x47\x53\x53\x41\x44\x00"

(* PNG signature (RFC 2083) — 0x89 is outside ASCII, hard-coded. *)
let magic_png = Bytes.of_string "\x89\x50\x4E\x47\x0D\x0A\x1A\x0A"
let magic_ogg = Bytes.of_string "OggS"
let magic_m4a = Bytes.of_string "ftyp"
let magic_riff = Bytes.of_string "RIFF"
let magic_webp = Bytes.of_string "WEBP"
let magic_jpg = Bytes.of_string "\xFF\xD8\xFF"

(** Equality of a fixed-size substring inside a larger buffer. *)
let sub_array_eq (offset : int) (length : int) (expected : bytes) (arr : bytes) :
    bool =
  if Bytes.length arr < offset + length then false
  else if Bytes.length expected < length then false
  else begin
    let ok = ref true in
    for i = 0 to length - 1 do
      if Bytes.get arr (offset + i) <> Bytes.get expected i then ok := false
    done;
    !ok
  end

(** Does [arr] start with the bytes of [prefix]? *)
let starts_with (prefix : bytes) (arr : bytes) : bool =
  if Bytes.length arr < Bytes.length prefix then false
  else sub_array_eq 0 (Bytes.length prefix) prefix arr

(** Was the ciphertext already plaintext? Caller passes the first 16 bytes. *)
let looks_like_plaintext (head : bytes) : bool =
  if Bytes.length head = 0 then false
  else
    starts_with magic_png head
    || starts_with magic_ogg head
    || starts_with magic_jpg head
    || (Bytes.length head >= 12 && starts_with magic_riff head
       && sub_array_eq 8 4 magic_webp head)
    || (Bytes.length head >= 8 && sub_array_eq 4 4 magic_m4a head)

(** Single hex character -> 0..15. Raises on bad char. *)
let hex_nibble (c : char) : int =
  match c with
  | '0' .. '9' -> Char.code c - Char.code '0'
  | 'a' .. 'f' -> Char.code c - Char.code 'a' + 10
  | _ -> invalid_arg (Printf.sprintf "non-hex char '%c'" c)

(** Decode a 32-char hex string into a 16-byte buffer. *)
let decode_hex_key (hex : string) : bytes =
  let cleaned = String.lowercase_ascii (String.trim hex) in
  if String.length cleaned <> 32 then
    invalid_arg
      (Printf.sprintf "encryption key must be 32 hex chars, got %d"
         (String.length cleaned));
  let b = Bytes.make 16 '\000' in
  for i = 0 to 15 do
    let hi = hex_nibble cleaned.[2 * i] in
    let lo = hex_nibble cleaned.[(2 * i) + 1] in
    Bytes.set b i (Char.chr ((hi lsl 4) lor lo))
  done;
  b

(** XOR [buf] against [key] cyclically. Symmetric — encrypt == decrypt. *)
let xor_transform (key : bytes) (buf : bytes) : bytes =
  if Bytes.length key = 0 then invalid_arg "XOR key must be non-empty";
  let n = Bytes.length buf in
  let out = Bytes.make n '\000' in
  let klen = Bytes.length key in
  for i = 0 to n - 1 do
    let v =
      Char.code (Bytes.get buf i) lxor Char.code (Bytes.get key (i mod klen))
    in
    Bytes.set out i (Char.chr v)
  done;
  out

let is_mv_magic_header (head : bytes) : bool =
  Bytes.length head >= Bytes.length magic_mv_header
  && starts_with magic_mv_header head

let is_mz_magic_header (head : bytes) : bool =
  Bytes.length head >= Bytes.length magic_mz_header
  && starts_with magic_mz_header head

let is_rgssad_magic (head : bytes) : bool =
  Bytes.length head >= Bytes.length magic_rgssad_prefix
  && starts_with magic_rgssad_prefix head

let is_zip_magic (head : bytes) : bool =
  Bytes.length head >= 4
  && Bytes.get head 0 = 'P'
  && Bytes.get head 1 = 'K'
  && Bytes.get head 2 = '\x03'
  && Bytes.get head 3 = '\x04'

let zero_fill (buf : bytes) : unit = Bytes.fill buf 0 (Bytes.length buf) '\000'
